File: ui_flet/gateways_panel.py
"""Gateways panel — Telegram / WhatsApp bot toggles."""

from __future__ import annotations

import flet as ft

from ui_flet.theme import TH


def build_gateways_panel(page: ft.Page, state, backend=None) -> ft.Container:
    """Panel for starting/stopping Telegram & WhatsApp gateways."""
    tg_status = ft.Text("Stopped", size=12, color=TH.dim)
    wa_status = ft.Text("Stopped", size=12, color=TH.dim)
    log_view = ft.ListView(controls=[], expand=True, spacing=4, auto_scroll=True)

    def _log(msg: str):
        log_view.controls.append(ft.Text(msg, size=11, color=TH.dim))
        if len(log_view.controls) > 200:
            log_view.controls = log_view.controls[-100:]
        page.update()

    async def _toggle_telegram(e):
        enabled = e.control.value
        if hasattr(state, "telegram_enabled"):
            state.telegram_enabled = enabled
        if enabled:
            tg_status.value = "Starting…"
            tg_status.color = TH.accent
            page.update()
            try:
                if backend and hasattr(backend, "start_gateways"):
                    await backend.start_gateways(telegram=True, whatsapp=False)
                tg_status.value = "Running"
                _log("Telegram gateway started")
            except Exception as exc:
                tg_status.value = f"Error: {exc}"
                tg_status.color = TH.error_c if hasattr(TH, "error_c") else "#ff4444"
                _log(f"Telegram error: {exc}")
        else:
            tg_status.value = "Stopped"
            tg_status.color = TH.dim
            _log("Telegram gateway stopped")
        page.update()

    async def _toggle_whatsapp(e):
        enabled = e.control.value
        if hasattr(state, "whatsapp_enabled"):
            state.whatsapp_enabled = enabled
        if enabled:
            wa_status.value = "Starting…"
            wa_status.color = TH.accent
            page.update()
            try:
                if backend and hasattr(backend, "start_gateways"):
                    await backend.start_gateways(telegram=False, whatsapp=True)
                wa_status.value = "Running"
                _log("WhatsApp gateway started")
            except Exception as exc:
                wa_status.value = f"Error: {exc}"
                wa_status.color = TH.error_c if hasattr(TH, "error_c") else "#ff4444"
                _log(f"WhatsApp error: {exc}")
        else:
            wa_status.value = "Stopped"
            wa_status.color = TH.dim
            _log("WhatsApp gateway stopped")
        page.update()

    tg_switch = ft.Switch(
        label="Telegram Bot",
        value=getattr(state, "telegram_enabled", False),
        on_change=_toggle_telegram,
    )
    wa_switch = ft.Switch(
        label="WhatsApp Bot",
        value=getattr(state, "whatsapp_enabled", False),
        on_change=_toggle_whatsapp,
    )

    return ft.Container(
        content=ft.Column(
            [
                ft.Text("Gateways", size=16, weight=ft.FontWeight.W_600, color=TH.text),
                ft.Text("Connect Zena to messaging platforms.", size=12, color=TH.dim),
                ft.Container(height=12),
                ft.Row([tg_switch, tg_status], spacing=12),
                ft.Row([wa_switch, wa_status], spacing=12),
                ft.Divider(color=TH.divider, height=16),
                ft.Text("Gateway Log", size=13, weight=ft.FontWeight.W_500, color=TH.text),
                log_view,
            ],
            expand=True,
            spacing=8,
        ),
        expand=True,
        padding=20,
    )

File: rust_output/src/gateways_panel.rs
/* /// Gateways panel — Telegram / WhatsApp bot toggles.

use anyhow::{Result, Context};
use crate::theme::{TH};

/// Panel for starting/stopping Telegram & WhatsApp gateways.
pub fn build_gateways_panel(page: ft::Page, state: String, backend: String) -> Result<ft::Container> {
    // Panel for starting/stopping Telegram & WhatsApp gateways.
    let mut tg_status = flet::... */;
    /* let mut wa_status = flet::... */;
    /* let mut log_view = flet::... */;
    let _log = |msg| {
        log_view.controls.push(/* flet::Text */(msg, /* size= */ 11, /* color= */ TH.dim));
        if log_view.controls.len() > 200 {
            log_view.controls = log_view.controls[-100..];
        }
        page.update();
    };
    let _toggle_telegram = |e| {
        let mut enabled = e.control.value;
        if /* hasattr(state, "telegram_enabled".to_string()) */ true {
            state::telegram_enabled = enabled;
        }
        if enabled {
            tg_status.value = "Starting…".to_string();
            tg_status.color = TH.accent;
            page.update();
            // try:
            {
                if (backend && /* hasattr(backend, "start_gateways".to_string()) */ true) {
                    backend.start_gateways(/* telegram= */ true, /* whatsapp= */ false).await;
                }
                tg_status.value = "Running".to_string();
                _log("Telegram gateway started".to_string());
            }
            // except Exception as exc:
        } else {
            tg_status.value = "Stopped".to_string();
            tg_status.color = TH.dim;
            _log("Telegram gateway stopped".to_string());
        }
        page.update();
    };
    let _toggle_whatsapp = |e| {
        let mut enabled = e.control.value;
        if /* hasattr(state, "whatsapp_enabled".to_string()) */ true {
            state::whatsapp_enabled = enabled;
        }
        if enabled {
            wa_status.value = "Starting…".to_string();
            wa_status.color = TH.accent;
            page.update();
            // try:
            {
                if (backend && /* hasattr(backend, "start_gateways".to_string()) */ true) {
                    backend.start_gateways(/* telegram= */ false, /* whatsapp= */ true).await;
                }
                wa_status.value = "Running".to_string();
                _log("WhatsApp gateway started".to_string());
            }
            // except Exception as exc:
        } else {
            wa_status.value = "Stopped".to_string();
            wa_status.color = TH.dim;
            /* _log("WhatsApp gateway stopped".to_string());
        }
        page.update();
    };
    let mut tg_switch = flet::... */;
    /* let mut wa_switch = flet::... */;
    Ok(/* flet::Container */(/* content= */ /* flet::Column */(vec![/* flet::Text */("Gateways".to_string(), /* size= */ 16, /* weight= */ /* flet::FontWeight.W_600 */, /* color= */ TH.text), /* flet::Text */("Connect Zena to messaging platforms.".to_string(), /* size= */ 12, /* color= */ TH.dim), /* flet::Container */(/* height= */ 12), /* flet::Row */(vec![tg_switch, tg_status], /* spacing= */ 12), /* flet::Row */(vec![wa_switch, wa_status], /* spacing= */ 12), /* flet::Divider */(/* color= */ TH.divider, /* height= */ 16), /* flet::Text */("Gateway Log".to_string(), /* size= */ 13, /* weight= */ /* flet::FontWeight.W_500 */, /* color= */ TH.text), log_view], /* expand= */ true, /* spacing= */ 8), /* expand= */ true, /* padding= */ 20))
}
